File: 3164. Find the Number of Good Pairs II/Solution.py
from collections import Counter
from typing import List


class Solution:
    def numberOfPairs(self, nums1: List[int], nums2: List[int], k: int) -> int:
        cnt1 = Counter(x // k for x in nums1 if x % k == 0)
        if not cnt1:
            return 0

        ans, lim = 0, max(cnt1.keys()) + 1
        for x, c in Counter(nums2).items():
            for y in range(x, lim, x):
                if y in cnt1:
                    ans += cnt1[y] * c
        return ans

File: 3164. Find the Number of Good Pairs II/Solution.cpp
#include <algorithm>
#include <ranges>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
public:
    long long numberOfPairs(vector<int>& nums1, vector<int>& nums2, int k)
    {
        unordered_map<int, long long> cnt1;
        for (int x : nums1)
            if (x % k == 0)
                ++cnt1[x / k];
        if (cnt1.empty())
            return 0;

        unordered_map<int, long long> cnt2;
        for (int x : nums2)
            ++cnt2[x];

        long long ans = 0, mx = ranges::max_element(cnt1)->first;
        for (const auto& [x, c] : cnt2)
            for (int y = x; y <= mx; y += x)
                if (cnt1.contains(y))
                    ans += cnt1[y] * c;
        return ans;
    }
};
